lib/MusicBrainz/Server/Controller/Root.pm

```diff
@@ -365,7 +365,11 @@ sub end : ActionClass('RenderView')
 {
     my ($self, $c) = @_;
 
-    return if exists $c->action->attributes->{Minimal};
+    my $attrs = $c->action->attributes;
+
+    $c->generate_csrf_token if exists $attrs->{CSRFToken};
```

Comment:
We now generate the token in `Controller::Root::end` (which is called before the template renders on every request), but only if `CSRFToken` is set on the action. Which I think is nicer than having to manually call `generate_csrf_token` at the end of every action.